Tools/build/check_warnings.py

```diff
@@ -0,0 +1,156 @@
+    warnings: list[dict],
+    files_with_expected_warnings: set[str],
+) -> int:
+    """
+    Returns failure status if there are no warnings in the list of warnings for a file
+    that is in the list of files with expected warnings
+    """
+
+    # Create set of files with warnings
+    files_with_warnings = set()
+    for warning in warnings:
+        locations = warning['locations']
+        for location in locations:
+            for key in ['caret', 'start', 'end']:
+                if key in location:
+                    file = location[key]['file']
+                    file = file.lstrip('./') # Remove leading curdir if present
+                    files_with_warnings.add(file)
+
+    unexpected_improvements = []
+    for file in files_with_expected_warnings:
+        if file not in files_with_warnings:
+            unexpected_improvements.append(file)
+
+    if unexpected_improvements:
+        print("Unexpected improvements:")
+        for file in unexpected_improvements:
+            print(file)
+        return 1
+
+    return 0
+
+def main(argv: list[str] | None = None) -> int:
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--compiler-output-file-path",
+        type=str,
+        required=True,
+        help="Path to the compiler output file"
+    )
+    parser.add_argument(
+        "--warning-ignore-file-path",
+        type=str,
+        required=True,
+        help="Path to the warning ignore file"
+    )
+    parser.add_argument(
+        "--fail-on-regression",
+        action="store_true",
+        default=False,
+        help="Flag to fail if new warnings are found"
+    )
+    parser.add_argument(
+        "--fail-on-improvement",
+        action="store_true",
+        default=False,
+        help="Flag to fail if files that were expected to have warnings have no warnings"
+    )
+
+    args = parser.parse_args(argv)
+
+    exit_code = 0
+
+    # Check that the compiler output file is a valid path
+    if not Path(args.compiler_output_file_path).is_file():
+        print(f"Compiler output file does not exist: {args.compiler_output_file_path}")
+        return 1
+
+    # Check that the warning ignore file is a valid path
+    if not Path(args.warning_ignore_file_path).is_file():
+        print(f"Warning ignore file does not exist: {args.warning_ignore_file_path}")
+        return 1
+
+    with Path(args.compiler_output_file_path).open(encoding="UTF-8") as f:
+        compiler_output_file_contents = f.read()
+
+    with Path(args.warning_ignore_file_path).open(encoding="UTF-8") as clean_files:
+        files_with_expected_warnings = {
+            file.strip()
+            for file in clean_files
+            if file.strip() and not file.startswith("#")
+        }
+
+    warnings = extract_warnings_from_compiler_output(compiler_output_file_contents)
+
+    status = get_unexpected_warnings(warnings, files_with_expected_warnings)
+    if args.fail_on_regression: exit_code |= status
```

Comment:
Since `status` can only be `1`, should we skip the `exit_code |=` and instead `return 1`? These should be put on their own line inside the `if` block as well.

Reply:
I wanted the get unexpected warnings and the get unexpected improvements to both run regardless if one fails. I can put the setting the status on a new line